File: 2022/swift/Sources/advent2022/Day12.swift
struct Day12 : Day {

    let sampleInput = [
        "Sabqponm",
        "abcryxxl",
        "accszExk",
        "acctuvwj",
        "abdefghi" ]

    func grid (_ input :[String]) -> [[Character]] { input.map { Array($0) }}

    struct Pos : Hashable {
        let row, col :Int
        func hash (into :inout Hasher) { into.combine(row) ; into.combine(col) }
        func get (_ grid :[[Character]]) -> Character { grid[row][col] }
        func onNeighbors (op :(Pos) -> Void) {
            for (dr, dc) in ndeltas { op(Pos(row: row+dr, col: col+dc)) }
        }
        static func == (a: Pos, b: Pos) -> Bool { return a.row == b.row && a.col == b.col }
    }
    static let ndeltas = [(-1, 0), (0, -1), (1, 0), (0, 1)]

    func indices (_ input :[[Character]]) -> [Pos] {
        input.indices.flatMap { row in input[row].indices.map { col in Pos(row: row, col: col) } }
    }

    func heightOf (_ c :Character) -> Int { c == "S" ? 0 : c == "E" ? heightOf("z") : Int(c.asciiValue!) - 97 }
    func height (_ elev :[[Character]], _ pos :Pos) -> Int { heightOf(pos.get(elev)) }

    func search (_ elev :[[Character]], _ start :Pos, _ end :Pos) -> Int? {
        var scan = [(start, 0)]
        var minDist = [start: 0]
        let gwidth = elev[0].count, gheight = elev.count
        while (scan.count > 0) {
            let (pos, dist) = scan.removeFirst()
            if pos == end { return dist }
            let h = height(elev, pos), ndist = dist+1
            pos.onNeighbors { npos in
                if npos.row < 0 || npos.col < 0 || npos.row >= gheight || npos.col >= gwidth { return }
                if height(elev, npos) - h > 1 { return }
                if let odist = minDist[npos], odist <= ndist { return }
                minDist[npos] = ndist
                scan.insertSorted((npos, ndist), { (a, b) in a.1 < b.1 })
            }
        }
        return nil
    }

    func pos (_ input :[[Character]], of c :Character) -> Pos {
        indices(input).first(where: { $0.get(input) == c }) ?? Pos(row: 0, col: 0)
    }

    func part1 () throws -> String {
        let elevs = grid(try readInput(12))
        return String(search(elevs, pos(elevs, of: "S"), pos(elevs, of: "E")) ?? 0)
    }

    func part2 () throws -> String {
        let elevs = grid(try readInput(12))
        let starts = indices(elevs).filter({ height(elevs, $0) == 0 }), end = pos(elevs, of: "E")
        return String(starts.compactMap({ search(elevs, $0, end) }).min() ?? 0)
    }
}

File: 2022/scala/src/main/scala/Day12.scala
object Day12 extends Day(12):

  type Grid = Seq[Array[Char]]
  case class Pos (row :Int, col :Int):
    def get (grid :Grid) = grid(row)(col)
    def onNeighbors (op :Pos => Unit) =
      for ((dr, dc) <- ndeltas) op(Pos(row+dr, col+dc))

  val ndeltas = Array((-1, 0), (0, -1), (1, 0), (0, 1))

  def indices (input :Grid) =
    for (row <- input.indices ; col <- input(row).indices) yield Pos(row, col)
  def pos (input :Grid, c :Char) = indices(input).find(_.get(input) == c).get

  def heightOf (c :Char) :Int = if c == 'S' then 0 else if c == 'E' then heightOf('z') else c - 97
  def height (elev :Grid, pos :Pos) = heightOf(pos.get(elev))

  def search (elev :Grid, start :Pos, end :Pos) :Option[Int] =
    var scan = collection.mutable.PriorityQueue((start, 0))(Ordering.by(-_._2))
    var minDist = Map(start -> 0)
    val gwidth = elev(0).size ; val gheight = elev.size
    while scan.size > 0 do
      val (pos, dist) = scan.dequeue()
      if pos == end then return Some(dist)
      val h = height(elev, pos) ; val ndist = dist+1
      pos.onNeighbors { npos =>
        if npos.row < 0 || npos.col < 0 || npos.row >= gheight || npos.col >= gwidth then {}
        else if height(elev, npos) - h > 1 then {}
        else if minDist.getOrElse(npos, ndist+1) <= ndist then {}
        else
          minDist += (npos -> ndist)
          scan += (npos -> ndist)
      }
    return None

  override def answer1 (input :Seq[String]) =
    val elevs = input.map(_.toArray)
    search(elevs, pos(elevs, 'S'), pos(elevs, 'E')).get

  override def answer2 (input :Seq[String]) =
    val elevs = input.map(_.toArray) ; val end = pos(elevs, 'E')
    val starts = indices(elevs).filter({ height(elevs, _) == 0 })
    starts.flatMap({ search(elevs, _, end) }).min
